dvc/stage/__init__.py

```diff
@@ -573,15 +573,21 @@ def run(
         if (self.cmd or self.is_import) and not self.frozen and not dry:
             self.remove_outs(ignore_remove=False, force=False)
 
-        if (not self.frozen and self.is_import) or self.is_partial_import:
+        if (
+            self.is_import and (not self.frozen or kwargs.get("pull"))
+        ) or self.is_partial_import:
             self._sync_import(dry, force, kwargs.get("jobs", None), no_download)
         elif not self.frozen and self.cmd:
             self._run_stage(dry, force, **kwargs)
-        else:
+        elif kwargs.get("pull"):
```

Comment:
Should not this come before you run the stage?

Reply:
This is for pulling outputs of unchanged stages. 

The follow-up P.R. adds a new flag that enables pulling dependencies  see https://github.com/iterative/dvc/pull/9437#discussion_r1189920252